### PR TITLE
MULE-15624: dbCreateStruct function should be able to resolve BLOB and CLOB types

### PR DESCRIPTION
**Pull request by fsgonz**
CLOB

### CODE REVIEW

File: modules/db/src/main/java/org/mule/module/db/internal/domain/connection/DefaultDbConnection.java [hunk 1]
```diff
@@ -34,6 +43,9 @@
  */
 public class DefaultDbConnection extends AbstractDbConnection
 {
+    protected transient Log logger = LogFactory.getLog(getClass());
```
Comment (elrodro83) [inline modules/db/src/main/java/org/mule/module/db/internal/domain/connection/DefaultDbConnection.java]: use slf4j

File: modules/db/src/main/java/org/mule/module/db/internal/domain/connection/DefaultDbConnection.java [hunk 2]
```diff
@@ -319,6 +331,14 @@ public Array createArrayOf(String typeName, Object[] elements) throws SQLExcepti
     @Override
     public Struct createStruct(String typeName, Object[] attributes) throws SQLException
     {
+        try
+        {
+            resolveLobs(typeName, attributes);
+        }
+        catch (SQLException e)
+        {
+            logger.warn("Unable to resolve lobs. Proceeding with original attributes.");
```
Comment (elrodro83) [inline modules/db/src/main/java/org/mule/module/db/internal/domain/connection/DefaultDbConnection.java]: should at least the message of the exception be logged?

File: modules/db/src/main/java/org/mule/module/db/internal/domain/connection/DefaultDbConnection.java [hunk 3]
```diff
@@ -34,6 +43,9 @@
  */
 public class DefaultDbConnection extends AbstractDbConnection
 {
+    protected transient Log logger = LogFactory.getLog(getClass());
+
+    public static final int ATTR_TYPE_NAME_INDEX = 6;
```
Comment (elrodro83) [inline modules/db/src/main/java/org/mule/module/db/internal/domain/connection/DefaultDbConnection.java]: is this constant among db engines? if so, is there no constant or an alternative way to get it in the JDBC apis?

File: modules/db/src/main/java/org/mule/module/db/internal/domain/connection/DefaultDbConnection.java [hunk 4]
```diff
@@ -363,5 +383,85 @@ public int getNetworkTimeout() throws SQLException
...
+    }
+    
+
+    protected void doResolveLobIn(Object[] attributes, int index, String dataTypeName) throws SQLException
+    {
+        if (dataTypeName.equals(BLOB_DB_TYPE.getName()))
+        {
+            attributes[index] = createBlob(attributes[index]);
+        }
+        else
```
Comment (elrodro83) [inline modules/db/src/main/java/org/mule/module/db/internal/domain/connection/DefaultDbConnection.java]: should check the data type for this too?

File: modules/db/src/main/java/org/mule/module/db/internal/domain/connection/DefaultDbConnection.java [hunk 5]
```diff
@@ -363,5 +383,85 @@ public int getNetworkTimeout() throws SQLException
...
+        {
+            blob.setBytes(1, toByteArray((InputStream) attribute));
+        }
+        else if (attribute instanceof String)
+        {
+            blob.setBytes(1, ((String) attribute).getBytes());
+        }
+        else
+        {
+            throw new IllegalArgumentException(format("Cannot create a %s from a value of type %s", Struct.class.getName(), attribute.getClass()));
```
Comment (elrodro83) [inline modules/db/src/main/java/org/mule/module/db/internal/domain/connection/DefaultDbConnection.java]: put quotes around the variable values

File: modules/db/src/main/java/org/mule/module/db/internal/domain/connection/DefaultDbConnection.java [hunk 6]
```diff
@@ -363,5 +383,85 @@ public int getNetworkTimeout() throws SQLException
...
+        {
+            clob.setString(1, (String) attribute);
+        }
+        else if (attribute instanceof InputStream)
+        {
+            clob.setString(1, IOUtils.toString((InputStream) attribute));
+        }
+        else
+        {
+            throw new IllegalArgumentException(format("Cannot create a %s from a value of type %s", Struct.class.getName(), attribute.getClass()));
```
Comment (elrodro83) [inline modules/db/src/main/java/org/mule/module/db/internal/domain/connection/DefaultDbConnection.java]: put quotes around the variable values

File: modules/db/src/main/java/org/mule/module/db/internal/domain/connection/OracleDbConnection.java [hunk 7]
```diff
@@ -77,4 +87,50 @@ private Method getCreateArrayOfMethod(Connection delegate)
...
+    {
+        Map<Integer, String> dataTypes = new HashMap<Integer, String>();
+
+        PreparedStatement ps = this.prepareStatement(QUERY_TYPE_ATTRS);
+        ps.setString(1, typeName);
+        ResultSet resultSet = null;
+
+        try
+        {
+            resultSet = ps.executeQuery();
```
Comment (elrodro83) [inline modules/db/src/main/java/org/mule/module/db/internal/domain/connection/OracleDbConnection.java]: doesnt ResultSet implement AutoClosable/Closable?

File: modules/db/src/main/java/org/mule/module/db/internal/domain/connection/DefaultDbConnection.java (hunk 3 above)
Comment (fsgonz) [reply modules/db/src/main/java/org/mule/module/db/internal/domain/connection/DefaultDbConnection.java]: It is the jdbc api

https://docs.oracle.com/javase/8/docs/api/java/sql/DatabaseMetaData.html#getAttributes-java.lang.String-java.lang.String-java.lang.String-java.lang.String-

https://docs.oracle.com/javase/7/docs/api/java/sql/DatabaseMetaData.html

But there is no constant in the jdbc api for that index, that I find. Every example I found uses the index without a constant (in a sense I find it reasonable as it is something related exclusively to the method).

File: modules/db/src/main/java/org/mule/module/db/internal/domain/connection/OracleDbConnection.java (hunk 7 above)
Comment (fsgonz) [reply modules/db/src/main/java/org/mule/module/db/internal/domain/connection/OracleDbConnection.java]: It is not needed from java 7.

File: modules/db/src/main/java/org/mule/module/db/internal/domain/connection/DefaultDbConnection.java [hunk 8]
```diff
@@ -337,7 +342,7 @@ public Struct createStruct(String typeName, Object[] attributes) throws SQLExcep
         }
         catch (SQLException e)
         {
-            logger.warn("Unable to resolve lobs. Proceeding with original attributes.");
+            logger.warn("Unable to resolve lobs. Proceeding with original attributes.", e);
```
Comment (elrodro83) [inline modules/db/src/main/java/org/mule/module/db/internal/domain/connection/DefaultDbConnection.java]: this will log the full stack trace, only the message of the exception is of interest

File: modules/db/src/main/java/org/mule/module/db/internal/domain/connection/DefaultDbConnection.java [hunk 9]
```diff
@@ -383,45 +388,44 @@ public int getNetworkTimeout() throws SQLException
...
-                String dataType = resultSet.getString(ATTR_TYPE_NAME_INDEX);
-                
-                doResolveLobIn(attributes, index, dataType);
-                
-                index++;
-            }
-        }
-        finally 
+
+        resultSet = this.getMetaData().getAttributes(this.getCatalog(), null, typeName, null);
```
Comment (elrodro83) [inline modules/db/src/main/java/org/mule/module/db/internal/domain/connection/DefaultDbConnection.java]: why did you remove the try? this won't close the ResultSet like this

File: modules/db/src/main/java/org/mule/module/db/internal/domain/connection/DefaultDbConnection.java [hunk 10]
```diff
@@ -383,45 +388,44 @@ public int getNetworkTimeout() throws SQLException
...
+
+    protected void doResolveLobIn(Object[] attributes, int index, int dataType, String dataTypeName) throws SQLException
     {
-        if (dataTypeName.equals(BLOB_DB_TYPE.getName()))
+        if (dataTypeName.equals(BLOB_DB_TYPE.getName()) && (dataType == UNKNOWN_DATA_TYPE || dataType == BLOB_DB_TYPE.getId()))
         {
             attributes[index] = createBlob(attributes[index]);
         }
-        else
+        else if (dataTypeName.equals(CLOB_DB_TYPE.getName()) && (dataType == UNKNOWN_DATA_TYPE || dataType == CLOB_DB_TYPE.getId()))
```
Comment (elrodro83) [inline modules/db/src/main/java/org/mule/module/db/internal/domain/connection/DefaultDbConnection.java]: extract the common condition between this and the previuos ti an outer if

File: modules/db/src/main/java/org/mule/module/db/internal/domain/connection/OracleDbConnection.java [hunk 11]
```diff
@@ -114,22 +120,13 @@ protected void resolveLobs(String typeName, Object[] attributes) throws SQLExcep
         ps.setString(1, typeName);
         ResultSet resultSet = null;
 
-        try
-        {
-            resultSet = ps.executeQuery();
+        resultSet = ps.executeQuery();
```
Comment (elrodro83) [inline modules/db/src/main/java/org/mule/module/db/internal/domain/connection/OracleDbConnection.java]: try-with-resources

File: modules/db/src/main/java/org/mule/module/db/internal/domain/connection/DefaultDbConnection.java [hunk 12]
```diff
@@ -342,7 +343,7 @@ public Struct createStruct(String typeName, Object[] attributes) throws SQLExcep
         }
         catch (SQLException e)
         {
-            logger.warn("Unable to resolve lobs. Proceeding with original attributes.", e);
+            logger.warn("Unable to resolve lobs. Proceeding with original attributes.", e.getMessage());
```
Comment (elrodro83) [inline modules/db/src/main/java/org/mule/module/db/internal/domain/connection/DefaultDbConnection.java]: missing the placeholder iun the message

File: modules/db/src/main/java/org/mule/module/db/internal/domain/connection/DefaultDbConnection.java [hunk 13]
```diff
@@ -319,6 +337,14 @@ public Array createArrayOf(String typeName, Object[] elements) throws SQLExcepti
     @Override
     public Struct createStruct(String typeName, Object[] attributes) throws SQLException
     {
+        try
+        {
+            resolveLobs(typeName, attributes);
+        }
+        catch (SQLException e)
+        {
+            logger.warn("Unable to resolve lobs. Proceeding with original attributes.", e.getMessage());
```
Comment (elrodro83) [inline modules/db/src/main/java/org/mule/module/db/internal/domain/connection/DefaultDbConnection.java]: missing placeholder in the message

File: modules/db/src/main/java/org/mule/module/db/internal/domain/connection/OracleDbConnection.java [hunk 14]
```diff
@@ -77,4 +87,49 @@ private Method getCreateArrayOfMethod(Connection delegate)
...
+    private Map<Integer, String> getDataTypes(String typeName) throws SQLException
+    {
+        Map<Integer, String> dataTypes = new HashMap<Integer, String>();
+
+        try (PreparedStatement ps = this.prepareStatement(QUERY_TYPE_ATTRS))
+        {
+            ps.setString(1, typeName);
+            ResultSet resultSet = null;
+
+            resultSet = ps.executeQuery();
```
Comment (elrodro83) [inline modules/db/src/main/java/org/mule/module/db/internal/domain/connection/OracleDbConnection.java]: inline

File: modules/db/src/main/java/org/mule/module/db/internal/domain/connection/DefaultDbConnection.java [hunk 15]
```diff
@@ -364,4 +390,96 @@ public int getNetworkTimeout() throws SQLException
...
+        }
+        else if (shouldResolveAttributeWithJdbcType(dataType, dataTypeName, CLOB_DB_TYPE))
+        {
+            attributes[index] = createClob(attributes[index]);
+        }
+    }
+
+    private boolean shouldResolveAttributeWithJdbcType(int dbDataType, String dbDataTypeName, DbType jdbcType)
+    {
+        if (dbDataType == UNKNOWN_DATA_TYPE || dbDataType == jdbcType.getId())
```
Comment (elrodro83) [inline modules/db/src/main/java/org/mule/module/db/internal/domain/connection/DefaultDbConnection.java]: join both ifs with an &&

File: modules/db/src/main/java/org/mule/module/db/internal/domain/connection/DefaultDbConnection.java (hunk 4 above)
Comment (fsgonz) [reply modules/db/src/main/java/org/mule/module/db/internal/domain/connection/DefaultDbConnection.java]: I've change this so that in case the dbDataType is known from the jdbc implementation, it is used to determine if the attribute has a type that has to be resolved. Otherwise the jdbc type name is used.
